### PR TITLE
FlappyBug: Introduce a new game

### PR DESCRIPTION
**Pull request by mimhufford**
This adds a new "Flappy Bug" style game to the collection :)

It has all the basic features.

Actual graphics and other quality of life improvements are planned, but I just wanted to get something going to start it off!

### CODE REVIEW

**Comment by gunnarbeutner** [comment]
When you hit a wall and hit a key right afterwards the score screen is dismissed and a new game is started. Maybe we should ignore key events for like a second or so so the user doesn't accidentally dismiss the score screen? :)

**Comment by mimhufford** [comment]
Good thinking! I'll add it to my to-do list :)

**Comment by gunnarbeutner** [comment]
This currently crashes for me on start:

```
ASSERTION FAILED: !PtrTraits::is_null(bits)
../.././AK/RefPtr.h:437

0x958821ce: [libsystem.so] syscall2 +0xe (syscall.cpp:25)
0xb32c7135: [libc.so] raise +0x25 (signal.cpp:21)
0xb32b02a6: [libc.so] abort +0x2a (stdlib.cpp:233)
0xb32b0e0b: [libc.so] __assertion_failed +0xeb (Format.h:236)
0x3493174f: [/bin/FlappyBug] FlappyBug::Game::paint_event(GUI::PaintEvent&) [clone .localalias] +0x70f (RefPtr.h:437)
0x5c9a24ec: [libgui.so] GUI::Widget::handle_paint_event(GUI::PaintEvent&) +0x48c (Widget.cpp:281)
0x12068bce: [libcore.so] Core::Object::dispatch_event(Core::Event&, Core::Object*) +0x7e (Object.h:113)
0x5c9b0d44: [libgui.so] GUI::Window::handle_multi_paint_event(GUI::MultiPaintEvent&) +0x2e4 (Window.cpp:406)
0x5c9b363a: [libgui.so] GUI::Window::event(Core::Event&) +0x18a (Window.cpp:547)
0x12068bce: [libcore.so] Core::Object::dispatch_event(Core::Event&, Core::Object*) +0x7e (Object.h:113)
0x1204f693: [libcore.so] Core::EventLoop::pump(Core::EventLoop::WaitMode) [clone .localalias] +0x4f3 (Atomic.h:246)
0x1204fe7b: [libcore.so] Core::EventLoop::exec() +0x5b (EventLoop.cpp:362)
0x5c89a0a8: [libgui.so] GUI::Application::exec() +0x28 (Application.cpp:106)
0x34930845: [/bin/FlappyBug] main +0x4f5 (NonnullRefPtr.h:276)
0x34930a08: [/bin/FlappyBug] _start +0x58 (crt0.cpp:39)
```

**Comment by mimhufford** [comment]
Hmm, I'll investigate tomorrow!

**Comment by mimhufford** [comment]
I've been able to repro and get the same backtrace, but only by manually entering incorrect paths to the images [here](https://github.com/SerenityOS/serenity/blob/2924e1fa144d3ba8763f585d6c0b6480cd7886ba/Userland/Games/FlappyBug/Game.h#L43). I made sure to rebase to be inline with upstream, but everything still appears to be working and the images used are definitely part of the commit. I'm a little lost to be honest so any help would be appreciated!

File: Userland/Games/FlappyBug/Game.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    void reset();
+    void game_over();
+    bool ready_to_start() const;
+
+    struct Bug {
+        const float x { 50 };
+        const float radius { 16 };
+        const float starting_y { 200 };
+        const RefPtr<Gfx::Bitmap> falling_bitmap { Gfx::Bitmap::load_from_file("res/icons/flappybug/falling.png") };
+        const RefPtr<Gfx::Bitmap> flapping_bitmap { Gfx::Bitmap::load_from_file("res/icons/flappybug/flapping.png") };
```
Comment (gunnarbeutner) [inline Userland/Games/FlappyBug/Game.h:44]:
```suggestion
        const RefPtr<Gfx::Bitmap> falling_bitmap { Gfx::Bitmap::load_from_file("/res/icons/flappybug/falling.png") };
        const RefPtr<Gfx::Bitmap> flapping_bitmap { Gfx::Bitmap::load_from_file("/res/icons/flappybug/flapping.png") };
```

You're using a relative path here which means this only works when executing the game with a current working dir of `/`.

Comment (mimhufford) [reply Userland/Games/FlappyBug/Game.h:44]: Amazing how hard you can face palm when working in a whole new environment :smile: thanks Gunnar!

Comment (gunnarbeutner) [reply Userland/Games/FlappyBug/Game.h]: You're welcome! I guess you wouldn't notice this when you usually launch the game through the taskbar menu. :)

File: Userland/Games/FlappyBug/Game.h [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        void apply_velocity()
+        {
+            y += velocity;
+        }
+    };
+
+    struct Obstacle {
+        const float width { 20 };
+        Color color { Color::DarkGray };
+        float x;
```
Comment (alimpfard) [inline Userland/Games/FlappyBug/Game.h:85]: Missing default initialiser.

File: Userland/Games/FlappyBug/Game.h [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        };
+        float x;
+        float y;
+        int bitmap_id;
+
+        Cloud()
+        {
+            bitmap_id = get_random_uniform(cloud_bitmaps.size());
+            x = get_random_uniform(game_width);
+            y = get_random_uniform(game_height / 2) + bitmap()->height();
```
Comment (alimpfard) [inline Userland/Games/FlappyBug/Game.h:120]: call `reset()` here to avoid duplicating this?

File: Userland/Games/FlappyBug/Game.h [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    };
+
+    Bug m_bug;
+    Obstacle m_obstacle;
+    Cloud m_cloud;
+    bool m_active;
+    Optional<float> m_highscore;
+    float m_last_score;
+    float m_difficulty;
+    float m_restart_cooldown;
```
Comment (alimpfard) [inline Userland/Games/FlappyBug/Game.h:148]: All of these primitive types are missing default initialisers.

**Review by alimpfard — COMMENTED** [review]
Just missing a few default initialisers, other stuff looks okay.

File: Userland/Games/FlappyBug/Game.h (hunk 3 above)
Comment (mimhufford) [reply Userland/Games/FlappyBug/Game.h]: I originally didn't do this as the `x` value ends up being slightly different in each case (`ctor` places the cloud on screen, `reset()` places it off-screen), but you're right :)

File: Userland/Games/FlappyBug/Game.h (hunk 4 above)
Comment (mimhufford) [reply Userland/Games/FlappyBug/Game.h]: Fixed, thanks!

**Comment by gunnarbeutner** [comment]
One thing I've noticed is you have several commits which fix issues that were introduced in this PR (e.g. `FlappyBug: Remove unused function`, `FlappyBug: Switch to absolute paths`, etc.). Can you squash those commits into whichever commit that added those issues? :)

**Comment by mimhufford** [comment]
> One thing I've noticed is you have several commits which fix issues that were introduced in this PR (e.g. `FlappyBug: Remove unused function`, `FlappyBug: Switch to absolute paths`, etc.). Can you squash those commits into whichever commit that added those issues? :)

Sorted, thanks for the comment! I'd never really done any git `squash`ing or `fixup`s before, nice to be forced into learning something new :smile:

**Pull request merged**